### PR TITLE
[action] notarize: add support for AppStore Connect API Key

### PR DESCRIPTION
**Pull request by niezbop**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Following Apple's incentives towards using AppStore Connect API keys instead of passwords, my team is trying to move to this new system. `notarize`, which is a critical component in our Mac builds pipeline does not seem to support it however, which is what this pull request aims to solve.

### Description

The biggest change in the interface of `notarize` this PR suggests is the addition of a new parameter, `api_key_path`, which would be the path to an API key file.

On a purely black box perspective, the only change there is the support for this new authentication method. I made this backwards compatibility in mind, so this change should not change anything for users that do not use API keys.

On a more technical level, I've done a few things:
- make notarize calls authentication abstract of which mean of authentication is used
- select authentication method based on the absence or presence of api_key_path
- support API key parameters
- I took the liberty of making `self.run` ignore rubocop Perceived Complexity metric, because addressing it would have me change more than what I was comfortable with for this specific PR

One important specificity of notarize (which I believe is not present in other apple tools) is that in order to be able to use the key, it must be present on the file system in a specific location, which means that I had to actually write the content of the token on the file system.

### CODE REVIEW

**Comment by niezbop** [comment]
@googlebot I signed it!

**Comment by lacostej** [comment]
@niezbop circleci fails probably require to rebase (See bbcf827)

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 1]
```diff
@@ -177,7 +213,15 @@ def self.available_options
                                        description: 'Whether to log requests',
                                        optional: true,
                                        default_value: false,
-                                       type: Boolean)
+                                       type: Boolean),
+          FastlaneCore::ConfigItem.new(key: :api_key_path,
+                                       env_name: 'FL_NOTARIZE_API_KEY_PATH',
+                                       description: 'Path to AppStore Connect API key',
+                                       optional: true,
+                                       is_string: true,
```
Comment (lacostej) [inline fastlane/lib/fastlane/actions/notarize.rb:221]: we could add `conflicting_options: [:username]` and vice versa.

**Review by lacostej — APPROVED** [review]
Looks good to me. Small non blocking improvements noted.
 
Would be good to get someone else to approve.

@joshdholtz maybe?

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 2]
```diff
@@ -123,6 +130,35 @@ def self.staple(package_path, verbose)
...
+          api_key = JSON.parse(File.read(api_key_path))
+          api_key_folder_path = '~/.appstoreconnect/private_keys'
+          api_key_file_path = File.join(api_key_folder_path, "AuthKey_#{api_key['key_id']}.p8")
+          directory_exists = File.directory?(api_key_folder_path)
+          file_exists = File.exist?(api_key_file_path)
+          begin
+            FileUtils.mkdir_p(api_key_folder_path) unless directory_exists
+            File.open(api_key_file_path, 'w') { |f| f.write(api_key['key']) } unless file_exists
+
+            yield(proc { |command| "#{command} --apiKey #{api_key['key_id']} --apiIssuer #{api_key['issuer_id']}" })
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/notarize.rb:146]: I’m thinking maybe we should do some validation `key` and `key_id` exist in `api_key` at the start of this method to make any debugging easier on the user if something isn’t right?

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Overall this looks great! A few small thoughts and I think implementing some of @lacostej suggestions would be 🔥 

And thanks fo making this PR ❤️

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 3]
```diff
@@ -123,6 +130,35 @@ def self.staple(package_path, verbose)
         )
       end
 
+      def self.with_notarize_authenticator(api_key_path)
+        if api_key_path
+          # From xcrun altool for --apiKey:
+          # This option will search the following directories in sequence for a private key file with the name of 'AuthKey_<api_key>.p8':  './private_keys', '~/private_keys', '~/.private_keys', and '~/.appstoreconnect/private_keys'.
+          api_key = JSON.parse(File.read(api_key_path))
```
[the remark this answers is not in the document]
Comment (lacostej) [reply fastlane/lib/fastlane/actions/notarize.rb:137]: not sure if smart. More like got bit by it once :)

I think I might have either opened an issue or a question about this somewhere already.

Found it: #18168 

If we reuse the `Token` class, we can fix that later.

File: fastlane/lib/fastlane/actions/notarize.rb (hunk 2 above)
Comment (lacostej) [reply fastlane/lib/fastlane/actions/notarize.rb:146]: or maybe we just reuse this 
https://github.com/fastlane/fastlane/blob/6702e2eeef1f383b3627b940f6e6c76d4ebc1551/spaceship/lib/spaceship/connect_api/token.rb#L28-L40

**Comment by niezbop** [comment]
@joshdholtz @lacostej Thanks for taking a look at this! I have leveraged `Spaceship::ConnectAPI::Token` to address most of your comments, and tackled the rest directly!

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 4]
```diff
@@ -123,6 +130,35 @@ def self.staple(package_path, verbose)
         )
       end
 
+      def self.with_notarize_authenticator(api_key_path)
+        if api_key_path
+          # From xcrun altool for --apiKey:
+          # This option will search the following directories in sequence for a private key file with the name of 'AuthKey_<api_key>.p8':  './private_keys', '~/private_keys', '~/.private_keys', and '~/.appstoreconnect/private_keys'.
```
[the remark this answers is not in the document]
Comment (niezbop) [reply fastlane/lib/fastlane/actions/notarize.rb:136]: Yep this is indeed a strange restriction that forces us to do some weird stuff that I'd rather have avoided.

File: fastlane/lib/fastlane/actions/notarize.rb (hunk 1 above)
Comment (niezbop) [reply fastlane/lib/fastlane/actions/notarize.rb:224]: Made optional and conflicting!

No logic is necessary to handle the case none of them is there, there will always be at least the `username` from its default value:
https://github.com/fastlane/fastlane/blob/afd4540ea9fdbff20813fc32acd60d6e7febf7bd/fastlane/lib/fastlane/actions/notarize.rb#L135-L136

**Comment by lacostej** [comment]
Looks good to me. Thanks @niezbop

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 5]
```diff
@@ -123,6 +130,35 @@ def self.staple(package_path, verbose)
         )
       end
 
+      def self.with_notarize_authenticator(api_key_path)
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/notarize.rb:133]:
```suggestion
      def self.with_notarize_authenticator(params, api_key_path)
```

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 6]
```diff
@@ -28,68 +30,73 @@ def self.run(params)
...
-        notarization_upload_command = "xcrun altool --notarize-app -t osx -f \"#{compressed_package_path || package_path}\" --primary-bundle-id #{bundle_id} -u #{apple_id_account.user} -p @env:FL_NOTARIZE_PASSWORD --output-format xml"
-        notarization_upload_command << " --asc-provider \"#{params[:asc_provider]}\"" if params[:asc_provider]
+        notarization_upload_command = "xcrun altool --notarize-app -t osx -f \"#{compressed_package_path || package_path}\" --primary-bundle-id #{bundle_id} --output-format xml"
 
-        notarization_upload_response = Actions.sh(
-          notarization_upload_command,
-          log: verbose
-        )
+        notarization_info = {}
+        with_notarize_authenticator(api_key_path) do |notarize_authenticator|
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/notarize.rb:38]:
```suggestion
        with_notarize_authenticator(params, api_key_path) do |notarize_authenticator|
```

File: fastlane/lib/fastlane/actions/notarize.rb [hunk 7]
```diff
@@ -123,6 +130,35 @@ def self.staple(package_path, verbose)
...
+          # From xcrun altool for --apiKey:
+          # This option will search the following directories in sequence for a private key file with the name of 'AuthKey_<api_key>.p8':  './private_keys', '~/private_keys', '~/.private_keys', and '~/.appstoreconnect/private_keys'.
+          api_key = Spaceship::ConnectAPI::Token.from_json_file(api_key_path)
+          api_key_folder_path = File.expand_path('~/.appstoreconnect/private_keys')
+          api_key_file_path = File.join(api_key_folder_path, "AuthKey_#{api_key.key_id}.p8")
+          directory_exists = File.directory?(api_key_folder_path)
+          file_exists = File.exist?(api_key_file_path)
+          begin
+            FileUtils.mkdir_p(api_key_folder_path) unless directory_exists
+            api_key.write_key_to_file(api_key_file_path) unless file_exists
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/notarize.rb:144]: I got errors when this action tried to write the file. I don’t _think_ I have my API file formatted incorrectly 🤔 

I _was_ able to get this action working by manually moving my `.p8` into this directory. Are you able to verify this still works on your end? 🙃 

```sh
Failed to authenticate with errors: (
    "Error Domain=ITunesConnectionAuthenticationErrorDomain Code=-26000 \"*** Failed to generate token: Error Domain=CupertinoJWT.CupertinoJWTError Code=1 \"The .p8 string has invalid format.\"\" UserInfo={NSLocalizedRecoverySuggestion=*** Failed to generate token: Error Domain=CupertinoJWT.CupertinoJWTError Code=1 \"The .p8 string has invalid format.\", NSLocalizedDescription=*** Failed to generate token: Error Domain=CupertinoJWT.CupertinoJWTError Code=1 \"The .p8 string has invalid format.\", NSLocalizedFailureReason=App Store operation failed.}"
)
```

**Review by joshdholtz — CHANGES_REQUESTED** [review]
Found one small issue where `params` needs to be passed with `with_notarize_authenticator`

Running into another issue where the `.p8` file isn’t getting written correctly. Can you verify this works on your end? 😊

File: spaceship/lib/spaceship/connect_api/token.rb [hunk 8]
```diff
@@ -97,6 +97,10 @@ def refresh!
       def expired?
         @expiration < Time.now
       end
+
+      def write_key_to_file(path)
+        File.open(path, 'w') { |f| f.write(@key) }
```
Comment (joshdholtz) [inline spaceship/lib/spaceship/connect_api/token.rb:102]: Ahhh, this is the reason its not working! The problem is that `@key` is a `OpenSSL::PKey::EC` object. The initializer of the `Token` class needs to be rewritten to take the private key text and keep that in memory before converting it to this object 🤔  You okay with making that change?

**Comment by joshdholtz** [comment]
@niezbop Let me know if you’d like me to make the changes I’ve requested 😊 I don’t have access to push a commit since this PR was made under an organization.  If so, I will need access to your fork. 😇

**Comment by niezbop** [comment]
Hey @joshdholtz ! I'm on holidays this week, so I won't be able to address these just now! If you want to tackle this yourself sooner, I (or @lacostej ) can add you as a collaborator on our fork 😄

**Comment by joshdholtz** [comment]
@niezbop Enjoy the holiday!! 

@lacostej Do you mind adding me as a collab? 😉

**Comment by niezbop** [comment]
@joshdholtz Looks like Jerome didn't give you access to the repo, so I made the changes!

Let me know if you think that's better.

**Review by joshdholtz — APPROVED** [review]
This is so good! Thanks for making those requested changes for me 😊 Really appreciate the time you put into this PR ❤️

**Comment by mokagio** [comment]
This is amazing! Thanks @niezbop for working on it! ❤️

**Comment by triplef** [comment]
This is really great, thank you @niezbop!

Just curious, was there any reason for this not using the data stored by the [app_store_connect_api_key](https://docs.fastlane.tools/actions/app_store_connect_api_key/) action in `SharedValues::APP_STORE_CONNECT_API_KEY`?

**Comment by joshdholtz** [comment]
@triplef Forgot about it 😇 Can fix!

**Comment by triplef** [comment]
> @triplef Forgot about it 😇 Can fix!

@joshdholtz I saw some changes re. App Store Connect API key in the last updates, but I couldn’t find info on whether this included support for notarize – would be great to see this if it’s not already added. 🙏

**Pull request merged**